File: Leetcode-solution-c-c++/reorderRoutesToMakeAllPathsLeadToTheCityZero.cpp
#include <bits/stdc++.h>
typedef long long int ll;
using namespace std;
class Solution
{
private:
    int dfs(int node, vector<unordered_set<int>> &unidirectional, vector<unordered_set<int>> &bidirectional, vector<bool> &visited)
    {
        int count = 0;
        for (int next : bidirectional[node])
        {
            if (visited[next])
                continue;

            if (unidirectional[node].count(next) == 0)
            {
                count++;
            }
            visited[next] = true;
            count += dfs(next, unidirectional, bidirectional, visited);
        }

        return count;
    }

public:
    int minReorder(int n, vector<vector<int>> &connections)
    {
        vector<unordered_set<int>> bidirectional(n, unordered_set<int>());
        vector<unordered_set<int>> unidirectional(n, unordered_set<int>());
        vector<bool> visited(n, false);
        visited[0] = true;

        for (auto connection : connections)
        {
            unidirectional[connection[1]].insert(connection[0]);
            bidirectional[connection[0]].insert(connection[1]);
            bidirectional[connection[1]].insert(connection[0]);
        }

        return dfs(0, unidirectional, bidirectional, visited);
    }
};

File: Leetcode-solution-python/reorderRoutesToMakeAllPathsLeadToTheCityZero.py
from typing import List, Set


class Solution:
    def dfs(self, node: int, bidirectional: List[List[int]], unidirectional: List[Set[int]],
            visited: List[bool]) -> int:
        count = 0

        for next_node in bidirectional[node]:
            if visited[next_node]:
                continue
            if node not in unidirectional[next_node]:
                count += 1
            visited[next_node] = True
            count += self.dfs(next_node, bidirectional, unidirectional, visited)

        return count

    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        bidirectional, unidirectional, visited = [list()] * n, [set()] * n, [False] * n
        visited[0] = True

        for a, b in connections:
            unidirectional[a].add(b)
            bidirectional[a].append(b)
            bidirectional[b].append(a)

        return self.dfs(0, bidirectional, unidirectional, visited)
